File: 33. Search in Rotated Sorted Array.cpp
class Solution {
public:
    int findIndex(vector< int > &nums, int low, int high , int lookFor) { 
        
        if( low > high ) return -1;
        int mid = ( low + high ) / 2;
        if(nums[mid] == lookFor ) return mid;
        else if( nums[mid] <= nums[high] ) {
            // case 1 
            if(nums[mid] < lookFor && lookFor <= nums[high] ) {
                return findIndex(nums, mid + 1 , high , lookFor );
            } else {
                return findIndex(nums , low , mid - 1 , lookFor );
            }
        } else {
            
           if(lookFor < nums[mid] && lookFor >= nums[low] ) {
                return findIndex(nums, low, mid - 1 , lookFor );
            }  else {
                return findIndex(nums, mid + 1 , high , lookFor );

            }


     }
        
    }
    int search(vector<int>& nums, int target) {
        int sz = nums.size();
        return findIndex(nums, 0, sz - 1 , target );
    }
};

File: python/Lakshay/33. Search in Rotated Sorted Array.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        low = 0
        high = len(nums) - 1
        def binarySearch(low, high):
            if low > high:
                return -1
            mid = low + ( high - low ) # 2 
            if nums[mid] == target:
                return mid
            else:
                if nums[low] <= nums[mid]:  # low to mid is sorted properly
                    if nums[low] <= target and target <= nums[mid]:
                        return binarySearch(low, mid - 1)
                    else:
                        return binarySearch( mid + 1, high )
                else:
                    if nums[mid] <= target and target <= nums[high]:
                        return binarySearch( mid + 1, high ) 
                    else:
                        return binarySearch( low, mid - 1 )
        
        return binarySearch( low, high )
                
